CASSANDRA-17226: Consistency Level Considered Connectivity Checker

frankgh opened the pull request:
This commit introduces two main features:

1. The ability to provide a custom implementation of the `StartupConnectivityChecker`
2. The `ConsistencyLevelStartupConnectivityChecker` checker

In `cassandra.yaml`, a custom implementation of the `StartupConnectivityChecker` can be
specified by providing the fully qualified class name by setting the `startup_connectivity_checker`
property in the yaml file. Additionally, the property supports the existing implementation
of the checker: `StrictStartupConnectivityChecker` (the default); and a new implementation
of the checker called the `ConsistencyLevelStartupConnectivityChecker`.

The current implementation of the `StartupConnectivityChecker` (the default) ensures that
all-but-one peers in each datacenter are available during Cassandra startup. We call this
implementation the strict implementation (`StrictStartupConnectivityChecker`). It is a
strict implemetation because it requires all-but-one of the peers are available during
cluster startup.

The Consistency Level Considered implementation of the checker borrows the concept of the
`ConsistencyLevel` to determine the number of peer connections required during startup.
The lowest non-zero replication factor among all user keyspaces is used to determine the
quorum.

The checker breaks the check into 2 phases:
1. In phase one, it checks if the primary set of peers has connected.
2. In phase two, if primary peers have not yet responded then it sends additional ping messages
   to the rest of the peers and checks if the connected peers can satisfy the specified
   consistency level.

The checker supports the following values for `ConsistencyLevel`:

- ALL: all peers are required for the checks
- QUORUM: primary peers required for quorum are used for the checks
- LOCAL_QUORUM: primary peers in the local DC required for local quorum are used for the checks
- EACH_QUORUM: quorum in each DC is required for the primary peers

The default consistency level is `LOCAL_QUORUM` and this value can be configured in
`cassandra.yaml` by specifying the `block_for_peers_consistency_level` property. The
checker reuses the original `block_for_peers_timeout_in_secs` property to determine
how long the node will wait to connect to other peers.

The cheker only supports the `NetworkTopologyStrategy`. The checker is not suitable
when using vnodes.

Co-authored-by: Yifan Cai <yifan_cai@apple.com>

clohfink commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -18,9 +18,21 @@
 
 import java.io.IOException;
-import java.net.*;
+import java.net.Inet4Address;
```
NP: maybe dont explode these out, next persons pr might fold them back up.

clohfink commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -1042,6 +1059,52 @@ public static void applySslContext()
+        // to work with StrictStartupClusterConnectivityChecker, so log a message if
+        // some other startup cluster connectivity checker is in use and the non-default
+        // value is detected
+        if (!(startupConnectivityChecker instanceof StrictStartupConnectivityChecker)
```
instead of tying this logic in DD, can add a `validate` method to `StartupConnectivityChecker` with the config allowing it to log warnings or fail. This also means any custom implementation can have checks.

frankgh replied:
This is warning the user that the properties specific for the `StrictStartupConnectivityChecker` are not unintentionally being configured with a different `startupConnectivityChecker`. So it's basically saying "hey, you changed the default value of this configuration property that is intended for a different startup checker than the one you are using". It's a pattern I saw somewhere else in this codebase, I will try to find where exactly that is being used.

frankgh replied:
so `validate` wouldn't make sense, since we are validating that other implementations are not using configuration properties intended for a different use case.

frankgh replied:
Found it: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/auth/AuthConfig.java#L63

clohfink replied:
I still think the StrictStartupConnectivityChecker validates could just confirm that the settings are not made for different checkers inside of it. But fair enough

clohfink commented on src/java/org/apache/cassandra/net/ConsistencyLevelStartupConnectivityChecker.java:
```diff
@@ -0,0 +1,442 @@
+ * <p>The checker only supports the {@link NetworkTopologyStrategy}. The checker is not suitable when using vnodes.
+ */
+public class ConsistencyLevelStartupConnectivityChecker
+extends AbstractStartupConnectivityChecker
```
NP: dont place implements/extends on new lines

clohfink commented on src/java/org/apache/cassandra/net/ConsistencyLevelStartupConnectivityChecker.java:
```diff
@@ -0,0 +1,442 @@
+ */
+public class ConsistencyLevelStartupConnectivityChecker
+extends AbstractStartupConnectivityChecker
+{
```
At a high level, I think a lot of this logic is replicated in the ReplicaPlan stuff that is used by the Read/Writes already. (hence will accurately reflect if they will throw UnavailableExceptions). Can we use this directly and cut almost all this logic? ie:

```            while (System.currentTimeMillis() < timeout)
            {
                boolean allGood = true;
                // different RFs
                for (Keyspace keyspace : Keyspace.all())
                {
                    for (TableMetadata table : keyspace.getMetadata().tables)
                    {
                        allGood &= RangeCommands.sufficientLiveNodesForSelectStar(table, consistencyLevel);
                    }
                }
                if (allGood) break;
                sleep(1000);
            }

frankgh replied:
Thanks for your review @clohfink. The drawback with this approach is that it does not establish small/large connections for the checks. Without those checks, we can only prevent unavailables and not timeouts.

frankgh committed:
CASSANDRA-17226: Consistency Level Considered Connectivity Checker

This commit introduces two main features:

1. The ability to provide a custom implementation of the `StartupConnectivityChecker`
2. The `ConsistencyLevelStartupConnectivityChecker` checker

In `cassandra.yaml`, a custom implementation of the `StartupConnectivityChecker` can be
specified by providing the fully qualified class name by setting the `startup_connectivity_checker`
property in the yaml file. Additionally, the property supports the existing implementation
of the checker: `StrictStartupConnectivityChecker` (the default); and a new implementation
of the checker called the `ConsistencyLevelStartupConnectivityChecker`.

The current implementation of the `StartupConnectivityChecker` (the default) ensures that
all-but-one peers in each datacenter are available during Cassandra startup. We call this
implementation the strict implementation (`StrictStartupConnectivityChecker`). It is a
strict implemetation because it requires all-but-one of the peers are available during
cluster startup.

The Consistency Level Considered implementation of the checker borrows the concept of the
`ConsistencyLevel` to determine the number of peer connections required during startup.
The lowest non-zero replication factor among all user keyspaces is used to determine the
quorum.

The checker breaks the check into 2 phases:
1. In phase one, it checks if the primary set of peers has connected.
2. In phase two, if primary peers have not yet responded then it sends additional ping messages
   to the rest of the peers and checks if the connected peers can satisfy the specified
   consistency level.

The checker supports the following values for `ConsistencyLevel`:

- ALL: all peers are required for the checks
- QUORUM: primary peers required for quorum are used for the checks
- LOCAL_QUORUM: primary peers in the local DC required for local quorum are used for the checks
- EACH_QUORUM: quorum in each DC is required for the primary peers

The default consistency level is `LOCAL_QUORUM` and this value can be configured in
`cassandra.yaml` by specifying the `block_for_peers_consistency_level` property. The
checker reuses the original `block_for_peers_timeout_in_secs` property to determine
how long the node will wait to connect to other peers.

The cheker only supports the `NetworkTopologyStrategy`. The checker is not suitable
when using vnodes.

Co-authored-by: Yifan Cai <yifan_cai@apple.com>

frankgh committed:
non-volatile

frankgh committed:
higher
